src/widget/whotcuebutton.cpp

```diff
@@ -129,16 +145,95 @@ void WHotcueButton::mousePressEvent(QMouseEvent* e) {
     }
 
     // Pass all other press events to the base class.
-    WPushButton::mousePressEvent(e);
+    WPushButton::mousePressEvent(pEvent);
 }
 
-void WHotcueButton::mouseReleaseEvent(QMouseEvent* e) {
-    const bool rightClick = e->button() == Qt::RightButton;
+void WHotcueButton::mouseReleaseEvent(QMouseEvent* pEvent) {
+    const bool rightClick = pEvent->button() == Qt::RightButton;
     if (rightClick) {
         // Don't handle stray release events
         return;
     }
-    WPushButton::mouseReleaseEvent(e);
+    WPushButton::mouseReleaseEvent(pEvent);
+}
+
+void WHotcueButton::mouseMoveEvent(QMouseEvent* pEvent) {
+    TrackPointer pTrack = PlayerInfo::instance().getTrackInfo(m_group);
+    if (!pTrack) {
+        return;
+    }
+    // Only allow moving set hotcues to empty or set slots.
+    // Note that Track::swapHotcues() allows both directions.
+    if (m_hotcue == Cue::kNoHotCue) {
+        return;
+    }
+
+    if (DragAndDropHelper::mouseMoveInitiatesDrag(pEvent)) {
+        QDrag* pDrag = new QDrag(this);
+        QMimeData* mimeData = new QMimeData;
+        mimeData->setText(mimeTextIdentifier() + kMimeTextDelimiter + QString::number(m_hotcue));
+        pDrag->setMimeData(mimeData);
+        // Use the currently rendered button as dnd cursor.
+        const QPixmap currLook = grab(rect().marginsRemoved(m_dndRectMargins));
+        pDrag->setDragCursor(currLook, Qt::MoveAction);
+        // Release this button after failed or successful drop.
+        // This prevents both the preview and the pressed state from getting stuck.
+        connect(pDrag,
+                &QDrag::destroyed,
+                this,
+                &WHotcueButton::release);
+        pDrag->exec();
+    }
+}
+
+void WHotcueButton::dragEnterEvent(QDragEnterEvent* pEvent) {
+    if (pEvent->source() == this) {
+        pEvent->ignore();
+        return;
+    }
+    TrackPointer pTrack = PlayerInfo::instance().getTrackInfo(m_group);
+    if (!pTrack) {
+        pEvent->ignore();
+        return;
+    }
+    const QString& mimeText = pEvent->mimeData()->text();
+    QStringList mimeTextLines = mimeText.split(kMimeTextDelimiter);
+    if (mimeTextLines.size() != 2 || mimeTextLines.at(0) != mimeTextIdentifier()) {
```

Comment:
since we previously essentially serialized the string data and now deserialize it again here, would it not make more sense instead to pack both in a struct and use binary data instead ([`QMimeData::setData`](https://doc.qt.io/qt-6/qmimedata.html#setData))?

Reply:
I'll look into that.

Reply:
does it need to be a `QString`?. What real information is encoded in there? From what I saw its just `"hotcue_"+m_group`, right? does it need to be `m_group` or can the same data be interpreted in a more binary-friendly fashion?

Reply:
Okay, we could as well use the TrackId (= DbId = int).

I came up with this c318193a58 (first time using QByteArray / QDataStream, so bare with me..)
But I'm not sure if the complexity is worth it.

Either way, using the TrackId instead of the player group is a good idea!

Reply:
yeah, thatll work. I thinks its the right direction, just needs a little cleanup

Reply:
feel free to push that commit so i can review it in context.